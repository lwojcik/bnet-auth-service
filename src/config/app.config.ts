import { registerAs } from '@nestjs/config';
import { trueStringToBoolean } from '../utils/trueStringToBoolean';
import { APP, APP_INFO, DEFAULTS } from '../common/constants';

const { env } = process;
const defaultValue = DEFAULTS.app;

export const appConfig = registerAs('app', () => ({
  name: APP_INFO.name,
  environment: env[APP.environment],
  host: env[APP.host] || defaultValue.host,
  port: parseInt(env[APP.port], 10) || defaultValue.port,
  enableCors:
    trueStringToBoolean({ value: env[APP.enableCors] }) ||
    defaultValue.enableCors,
  corsOrigin: env[APP.corsOrigin] || defaultValue.corsOrigin,
}));
