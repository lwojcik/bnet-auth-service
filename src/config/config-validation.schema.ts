import * as Joi from 'joi';
import { CronExpression } from '@nestjs/schedule';
import { validateRegionName } from '../common/validators/validateRegionName.validator';

const appSchema = {
  NODE_ENV: Joi.string().default('production'),
  BAS_APP_HOST: Joi.string().default('0.0.0.0'),
  BAS_APP_PORT: Joi.string().default(3000),
  BAS_APP_CORS_ORIGIN: Joi.optional(),
};

const redisSchema = {
  BAS_REDIS_HOST: Joi.string().optional().default('redis'),
  BAS_REDIS_PORT: Joi.string().optional().default('6379'),
  BAS_REDIS_PASSWORD: Joi.any().optional(),
  BAS_REDIS_TTL_SECS: Joi.string().optional().default('2000'),
  BAS_REDIS_DB: Joi.string().optional().default('0'),
};

const battlenetSchema = {
  BAS_BATTLENET_REGION: Joi.string().required().custom(validateRegionName),
  BAS_BATTLENET_CLIENT_ID: Joi.string().required(),
  BAS_BATTLENET_CLIENT_SECRET: Joi.string().required(),
};

const throttleSchema = {
  BAS_THROTTLE_LIMIT: Joi.string().optional().default('300'),
  BAS_THROTTLE_TTL_SECS: Joi.string().optional().default('60'),
};

const authSchema = {
  BAS_AUTH_USERNAME: Joi.string().optional(),
  BAS_AUTH_JWT_SECRET: Joi.string().optional(),
};

const httpsSchema = {
  BAS_HTTPS_KEY_PATH: Joi.string().optional(),
  BAS_HTTPS_CERT_PATH: Joi.string().optional(),
};

const cronSchema = {
  BAS_CRON_PATTERN: Joi.string()
    .optional()
    .default(CronExpression.EVERY_30_MINUTES),
};

export const configValidationSchema = Joi.object({
  ...appSchema,
  ...redisSchema,
  ...battlenetSchema,
  ...throttleSchema,
  ...authSchema,
  ...httpsSchema,
  ...cronSchema,
});
