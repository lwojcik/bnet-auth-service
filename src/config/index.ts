export * from './app.config';
export * from './auth.config';
export * from './battlenet.config';
export * from './endpoints.config';
export * from './redis.config';
export * from './throttle.config';
