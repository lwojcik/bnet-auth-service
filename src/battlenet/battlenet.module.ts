import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { LoggerModule } from '../logger/logger.module';
import { battleNetConfig } from '../config';
import { BattleNetService } from './battlenet.service';

@Module({
  imports: [ConfigModule.forFeature(battleNetConfig), LoggerModule],
  providers: [BattleNetService],
  exports: [BattleNetService],
})
export class BattleNetModule {}
