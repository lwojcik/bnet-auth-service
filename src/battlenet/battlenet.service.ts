import { Inject, Injectable } from '@nestjs/common';
import { ConfigType } from '@nestjs/config';
import { BlizzAPI, RegionName } from 'blizzapi';
import { ApiErrorCode } from '../common/types';
import { battleNetConfig } from '../config';
import { LoggerService } from '../logger/logger.service';

@Injectable()
export class BattleNetService {
  private readonly blizzApi: BlizzAPI;

  constructor(
    @Inject(battleNetConfig.KEY)
    private readonly bnetConfig: ConfigType<typeof battleNetConfig>,
    private readonly logger: LoggerService
  ) {
    this.blizzApi = new BlizzAPI({
      region: this.bnetConfig.region as RegionName,
      clientId: this.bnetConfig.clientId,
      clientSecret: this.bnetConfig.clientSecret,
    });
    this.logger.setLoggedClass(BattleNetService.name);
  }

  async getAccessToken() {
    this.logger.setLoggedMethod(this.getAccessToken.name);
    this.logger.debug();

    try {
      this.logger.debug(
        `Using Battle.net client id: ${this.bnetConfig.clientId}`
      );
      this.logger.debug(
        `Using Battle.net client secret:${this.bnetConfig.clientSecret}`
      );

      const accessToken = await this.blizzApi.getAccessToken();

      this.logger.debug(`Received access token: ${accessToken}`);

      return accessToken;
    } catch (error) {
      this.logger.error(error);

      return {
        error: ApiErrorCode.bnetApiError,
        statusCode: error.response.status,
        message: error.message,
      };
    }
  }
}
