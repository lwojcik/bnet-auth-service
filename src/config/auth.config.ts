import { registerAs } from '@nestjs/config';
import { trueStringToBoolean } from '../utils/trueStringToBoolean';
import { AUTH, DEFAULTS } from '../common/constants';

const { env } = process;
const defaultValue = DEFAULTS.auth;

export const authConfig = registerAs('auth', () => ({
  enable:
    trueStringToBoolean({ value: process.env[AUTH.enable] }) ||
    defaultValue.enable,
  username: env[AUTH.username] || defaultValue.username,
  jwtSecret: env[AUTH.jwtSecret] || defaultValue.jwtSecret,
}));
