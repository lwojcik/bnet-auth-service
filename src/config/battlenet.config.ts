import { registerAs } from '@nestjs/config';
import { BATTLENET } from '../common/constants';

const { env } = process;

export const battleNetConfig = registerAs('battlenet', () => ({
  region: env[BATTLENET.region],
  clientId: env[BATTLENET.clientId],
  clientSecret: env[BATTLENET.clientSecret],
}));
