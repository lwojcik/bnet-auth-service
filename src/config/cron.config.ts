import { registerAs } from '@nestjs/config';

export const cronConfig = registerAs('cron', () => ({
  enable: process.env.BAS_CRON_ENABLE === 'true',
  pattern: process.env.BAS_CRON_PATTERN,
}));
