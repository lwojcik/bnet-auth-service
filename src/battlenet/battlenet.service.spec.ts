import { Test, TestingModule } from '@nestjs/testing';
import { ConfigModule, registerAs } from '@nestjs/config';
import { BlizzAPI } from 'blizzapi';
import { LoggerService } from '../logger/logger.service';
import { BattleNetService } from './battlenet.service';

jest.mock('blizzapi');

describe('BattleNetService', () => {
  let service: BattleNetService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forFeature(
          registerAs('battlenet', () => ({
            region: 'sample-battlenet-region',
            clientId: 'sample-client-id',
            clientSecret: 'sample-client-secret',
          }))
        ),
      ],
      providers: [
        {
          provide: LoggerService,
          useValue: {
            debug: jest.fn(),
            log: jest.fn(),
            error: jest.fn(),
            setLoggedClass: jest.fn(),
            setLoggedMethod: jest.fn(),
          },
        },
        BattleNetService,
      ],
    }).compile();

    service = module.get<BattleNetService>(BattleNetService);
  });

  afterAll(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (BlizzAPI as any).mockClear();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should get access token', async () => {
    BlizzAPI.prototype.getAccessToken = () =>
      Promise.resolve('sample_access_token_from_blizzapi_mock');

    expect.assertions(1);
    const accessToken = await service.getAccessToken();
    expect(accessToken).toMatchSnapshot();
  });

  it('should return error object in case of Battle.net API error', async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (BlizzAPI as any).prototype.getAccessToken = () =>
      // eslint-disable-next-line prefer-promise-reject-errors
      Promise.reject({
        message: 'Sample error message from BlizzAPI mock',
        response: {
          status: 'Sample error status from BlizzAPI mock',
        },
      });

    expect.assertions(1);
    const accessToken = await service.getAccessToken();
    expect(accessToken).toMatchSnapshot();
  });
});
