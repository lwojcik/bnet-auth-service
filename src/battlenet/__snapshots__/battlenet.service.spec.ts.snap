// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BattleNetService should get access token 1`] = `"sample_access_token_from_blizzapi_mock"`;

exports[`BattleNetService should return error object in case of Battle.net API error 1`] = `
Object {
  "error": "BnetApiError",
  "message": "Sample error message from BlizzAPI mock",
  "statusCode": "Sample error status from BlizzAPI mock",
}
`;
